solidity/contracts/isms/hook/PolygonZkevmV2Ism.sol

```diff
@@ -0,0 +1,212 @@
+            uint32 globalIndex,
+            bytes32 mainnetExitRoot,
+            bytes32 rollupExitRoot,
+            uint32 originNetwork,
+            address originAddress,
+            ,
+            ,
+            uint256 amount,
+            bytes memory payload
+        ) = abi.decode(
+                _metadata,
+                (
+                    bytes32[32],
+                    bytes32[32],
+                    uint32,
+                    bytes32,
+                    bytes32,
+                    uint32,
+                    address,
+                    uint32,
+                    address,
+                    uint256,
+                    bytes
+                )
+            );
+
+        require(
+            messageId == abi.decode(payload, (bytes32)),
+            "PolygonZkevmV2Ism: message id does not match payload"
+        );
+        zkEvmBridge.claimMessage(
+            smtProofLocalExitRoot,
+            smtProofRollupExitRoot,
+            globalIndex,
+            mainnetExitRoot,
+            rollupExitRoot,
+            originNetwork,
+            originAddress,
+            zkEvmBridgeDestinationNetId,
+            address(this),
+            amount,
+            payload
+        );
+        uint256 _msgValue = verifiedMessages[messageId].clearBit(
+            VERIFIED_MASK_INDEX
+        );
+        if (_msgValue > 0) {
+            verifiedMessages[messageId] -= _msgValue;
+            payable(_message.recipientAddress()).sendValue(_msgValue);
+        }
+
+        return true;
+    }
+
+    /**
+     * @dev Callback function for Zkevm bridge.
+     * Verifies the received message.
+     * @inheritdoc IBridgeMessageReceiver
+     */
+    function onMessageReceived(
+        address,
+        uint32,
+        bytes memory data
+    ) external payable override {
+        require(
+            msg.sender == address(zkEvmBridge),
+            "PolygonZkevmV2Ism: invalid sender"
+        );
+        require(data.length == 32, "PolygonZkevmV2Ism: data must be 32 bytes");
+        require(
+            _isAuthorized(),
+            "AbstractMessageIdAuthorizedIsm: sender is not the hook"
+        );
+        require(
+            msg.value < 2 ** VERIFIED_MASK_INDEX,
+            "AbstractMessageIdAuthorizedIsm: msg.value must be less than 2^255"
+        );
+
+        bytes32 messageId = abi.decode(data, (bytes32));
+        bool verified = verifiedMessages[messageId].isBitSet(
+            VERIFIED_MASK_INDEX
+        );
+        require(!verified, "PolygonZkevmV2Ism: message already verified");
+        verifiedMessages[messageId] = msg.value.setBit(VERIFIED_MASK_INDEX);
+
+        emit ReceivedMessage(messageId);
+    }
+
+    /**
+     * @dev Checks if the origin chain message sender is the hook address.
+     * @inheritdoc AbstractMessageIdAuthorizedIsm
+     */
+    function _isAuthorized() internal view override returns (bool) {
+        bytes32 originSender = abi.decode(msg.data[4:], (bytes32));
```

Comment:
don't we need to access control msg.sender since msg.data[4:] can be arbitrarily set by the caller?